src/bos/common/clients/pcs/transitions.py

```diff
@@ -0,0 +1,149 @@
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+import logging
+
+from .base import BasePcsEndpoint
+from .exceptions import PowerControlComponentsEmptyException, PowerControlSyntaxException
+
+LOGGER = logging.getLogger(__name__)
+
+
+class TransitionsEndpoint(BasePcsEndpoint):
+    ENDPOINT = 'transitions'
+
+    def transition_create(self,
+                          xnames,
+                          operation,
+                          task_deadline_minutes=None,
+                          deputy_key=None):
+        """
+        Interact with PCS to create a request to transition one or more xnames. The transition
+        operation indicates what the desired operation should be, which is a string value
+        containing one or more of the supported transition names for the given hardware, e.g. 'on',
+        'off', or 'force-off'.
+
+        Once created, one of two responses are returned. A 2XX response results in a
+        transition_start_output object, or, an invalid request results in a 4XX and subsequent
+        raised PCS exception.
+
+        Args:
+            xnames: an iterable of xnames
+            operation: A string/enum for what the nodes should transition to
+            task_deadline_minutes: How long should PCS operate on the nodes to bring them to
+              complete the operation; typecast to an integer value.
+            deputy_key: An optional string value that can be used to further handle instructing PCS
+              to perform state transitions on behalf of a known existing reservation.
+            session: An already existing session to use with PCS, if any
+
+        Returns:
+            A transition_start_output object, which is a record for the transition that was
+            created. The most important field of which is the 'transitionID' value, which allows
+            subsequent follow-on to the created request.
+
+        Raises:
+            PowerControlException: Any non-nominal response from PCS, typically as a result of an
+              unexpected payload response, or a failure to create a transition record.
+            PowerControlComponentsEmptyException: No xnames specified
+        """
+        if not xnames:
+            raise PowerControlComponentsEmptyException(
+                f"_transition_create called with no xnames! (operation={operation})"
+            )
+        try:
+            assert operation in {
+                'On', 'Off', 'Soft-Off', 'Soft-Restart', 'Hard-Restart',
```

Comment:
better to use constants for each operation and put them in constants file.

Reply:
That part of the function is just copied from the previous PCS client. The task of moving the clients over to the new framework was big enough that I didn't look into ways to improve their existing logic. We can consider this for a different PR.